jax/random.py

```diff
@@ -878,18 +880,46 @@ def _case4(zagf):
 def _gamma_grad(sample, a):
     samples = np.reshape(sample, -1)
     alphas = np.reshape(a, -1)
-    grads = vmap(_gamma_grad_one)(samples, alphas)
+    if np.size(alphas) == 1:
+        grads = _gamma_grad_one(samples[0], alphas[0])
+    else:
+        # TODO: benchmark execute time against grads = vmap(_gamma_grad_one)(samples, alphas)
+        grads = lax.map(lambda args: _gamma_grad_one(*args), (samples, alphas))
     return grads.reshape(onp.shape(a))
 
-@custom_transforms
 def _gamma_impl(key, a):
-    alphas = np.reshape(a, -1)
-    keys = split(key, onp.size(alphas))
-    samples = vmap(_gamma_one)(keys, alphas)
-    return np.reshape(samples, onp.shape(a))
-
-defjvp(_gamma_impl, None,
-       lambda tangent, ans, key, a, **kwargs: tangent * _gamma_grad(ans, a))
+  if key.ndim == 2:  # batch of keys and alphas
+    size = np.size(a[0])
+    if size > 1:
+      key = lax.map(lambda k: split(k, size), key)
+  else:
+    size = np.size(a)
+    if size > 1:
+      key = split(key, size)
+  alphas = np.reshape(a, -1)
+  keys = np.reshape(key, (-1, 2))
+  if np.size(alphas) == 1:
+    samples = _gamma_one(keys[0], alphas[0])
+  else:
+    # XXX in GPU, using lax.map is slower than using vmap if alphas.size > 50000
+    # but that usage case is rare and can be resolved by vectorizing gamma sampler
+    samples = lax.map(lambda args: _gamma_one(*args), (keys, alphas))
+  return np.reshape(samples, np.shape(a))
+
+def _gamma_batching_rule(batched_args, batch_dims):
+    k, a = batched_args
+    bk, ba = batch_dims
+    size = next(t.shape[i] for t, i in zip(batched_args, batch_dims) if i is not None)
+    k = batching.bdim_at_front(k, bk, size)
+    a = batching.bdim_at_front(a, ba, size)
+    return random_gamma_p.bind(k, a), 0
+
+random_gamma_p = core.Primitive('random_gamma')
+random_gamma_p.def_impl(_gamma_impl)  # partial(xla.apply_primitive, random_gamma_p))
```

Comment:
I believe I should use `partial(xla.apply_primitive, random_gamma_p))` here but it doesn't work.